### PR TITLE
[Core] Refactor activation and normalization layers

### PR DESCRIPTION
**Pull request by sayakpaul**
Internal Slack thread (internal to Diffusers maintainers):
https://huggingface.slack.com/archives/C04DBJ0HLH1/p1697859469537249

@stevhliu docs look okay to you?

### CODE REVIEW

File: docs/source/en/_toctree.yml [hunk 1]
```diff
@@ -162,6 +162,10 @@
   title: Conceptual Guides
 - sections:
   - sections:
+    - local: api/activations
+      title: Custom Activation Functions
```
Comment (stevhliu) [inline docs/source/en/_toctree.yml:166]:
```suggestion
      title: Custom activation functions
```

File: docs/source/en/_toctree.yml [hunk 2]
```diff
@@ -162,6 +162,10 @@
...
+    - local: api/normalization
+      title: Custom Normalization Layers
```
Comment (stevhliu) [inline docs/source/en/_toctree.yml:168]:
```suggestion
      title: Custom normalization layers
```

File: src/diffusers/models/activations.py [hunk 3]
```diff
@@ -20,3 +40,76 @@ def get_activation(act_fn: str) -> nn.Module:
...
+
+    def forward(self, hidden_states):
+        hidden_states = self.proj(hidden_states)
+        hidden_states = self.gelu(hidden_states)
+        return hidden_states
+
+
+class GEGLU(nn.Module):
+    r"""
+    A variant of the gated linear unit activation function from https://arxiv.org/abs/2002.05202.
```
Comment (stevhliu) [inline src/diffusers/models/activations.py:74]:
```suggestion
    A [variant](https://arxiv.org/abs/2002.05202) of the gated linear unit activation function.
```

File: src/diffusers/models/activations.py [hunk 4]
```diff
@@ -20,3 +40,76 @@ def get_activation(act_fn: str) -> nn.Module:
...
+    def forward(self, hidden_states, scale: float = 1.0):
+        args = () if USE_PEFT_BACKEND else (scale,)
+        hidden_states, gate = self.proj(hidden_states, *args).chunk(2, dim=-1)
+        return hidden_states * self.gelu(gate)
+
+
+class ApproximateGELU(nn.Module):
+    r"""
+    The approximate form of Gaussian Error Linear Unit (GELU). For more details, see section 2:
+    https://arxiv.org/abs/1606.08415.
```
Comment (stevhliu) [inline src/diffusers/models/activations.py:102]:
```suggestion
    The approximate form of the Gaussian Error Linear Unit (GELU). For more details, see section 2 of
    this [paper](https://arxiv.org/abs/1606.08415).
```

File: src/diffusers/models/normalization.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+from .embeddings import CombinedTimestepLabelEmbeddings
+
+
+class AdaLayerNorm(nn.Module):
+    r"""
+    Norm layer modified to incorporate timestep embeddings.
+
+    Parameters:
+        embedding_dim (`int`): The size of each embedding vector.
+        num_embeddings (`int`): The size of the dictionary of embeddings.
```
Comment (stevhliu) [inline src/diffusers/models/normalization.py:32]:
```suggestion
        num_embeddings (`int`): The size of the embeddings dictionary.
```

File: src/diffusers/models/normalization.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        return x
+
+
+class AdaLayerNormZero(nn.Module):
+    r"""
+    Norm layer adaptive layer norm zero (adaLN-Zero).
+
+    Parameters:
+        embedding_dim (`int`): The size of each embedding vector.
+        num_embeddings (`int`): The size of the dictionary of embeddings.
```
Comment (stevhliu) [inline src/diffusers/models/normalization.py:55]:
```suggestion
        num_embeddings (`int`): The size of the embeddings dictionary.
```

File: src/diffusers/models/normalization.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        return x, gate_msa, shift_mlp, scale_mlp, gate_mlp
+
+
+class AdaGroupNorm(nn.Module):
+    r"""
+    GroupNorm layer modified to incorporate timestep embeddings.
+
+    Parameters:
+        embedding_dim (`int`): The size of each embedding vector.
+        num_embeddings (`int`): The size of the dictionary of embeddings.
```
Comment (stevhliu) [inline src/diffusers/models/normalization.py:86]:
```suggestion
        num_embeddings (`int`): The size of the embeddings dictionary.
```

**Review by stevhliu — APPROVED** [review]
Nice and clean docstrings! 👍

File: docs/source/en/_toctree.yml [hunk 8]
```diff
@@ -162,6 +162,10 @@
   title: Conceptual Guides
 - sections:
   - sections:
+    - local: api/activations
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply docs/source/en/_toctree.yml:165]: See: https://github.com/huggingface/transformers/tree/main/docs/source/en/internal

**Review by patrickvonplaten — COMMENTED** [review]
Looks good! Thanks for the clean-up. Let's just make sure that the classes are clearly marked as internal classes in the doc

File: docs/source/en/_toctree.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply docs/source/en/_toctree.yml:165]: Would it not make sense to update in a bulk PR? 

Cc: @stevhliu too.

[the remark this answers is not in the document]
Comment (stevhliu) [reply docs/source/en/_toctree.yml:165]: Sure, I can update and create a new "internal classes" section in a separate PR if you'd like!

[the remark this answers is not in the document]
Comment (sayakpaul) [reply docs/source/en/_toctree.yml:165]: That'd be great. So, I am merging it now then.
